Added training with custom data support, fixed bugs with ddpm

wu-ming233 opened the pull request:
This pull request entails:
- Streamlining training on new data directly from midi files using an updated dataloader
- Supporting an arbitrary amount of tracks in the input, as opposed to the fixed (melody, bridge, piano) shape from POP909
- Extending the chord extractor with the option to round the duration of chords to the nearest beat. This significantly reduces error rate in preparing training data, while maintaining a reasonable accuracy
- Fixing `collate_fn` shifting unevenly in dataloader
- Fixing vanilla ddpm model not being able to train with `PolyffusionLearner`
- Fixing ddpm not being able to inference when `length=1`
- Enabling retraining from checkpoint with an increased `max_epochs`
- Implementing several convenient changes for batch inferencing

These changes are mainly just small additions to enhance user experience, and are completely backward-compatible.

wu-ming233 committed:
Added training with custom data support, fixed bugs with ddpm, and improved batch inferencing

aik2mlj commented on polyffusion/data/dataloader.py:
```diff
@@ -19,7 +19,7 @@
 def collate_fn(batch, shift):
     def sample_shift():
-        return np.random.choice(np.arange(-6, 6), 1)[0]
+        return np.random.choice(np.arange(-6, 7), 1)[0]
```
This seems weird to me. We randomly transpose the key to 12 tones evenly. But now it is randomized to 13 tones. I think the former is correct.

wu-ming233 replied:
In numpy's documentation, `np.arange(start, stop)` generates within the half-open interval `[start, stop)`. Thus, to generate 12 tones evenly, I believe we would need (-6, 7). See https://numpy.org/doc/stable/reference/generated/numpy.arange.html

aik2mlj replied:
Well I know, but [-6, 6) is transposing 12 tones evenly. If it's [-6, 7), the -6 transposed and the +6 transposed versions are on the same key.

wu-ming233 replied:
My mistake. Sorry about that! Commit added

aik2mlj commented on polyffusion/main.py:
Let's add `data_dir=args.data_dir` to all the model configs!

wu-ming233 replied:
Agreed!

aik2mlj commented on polyffusion/prepare_data.py:
```diff
@@ -0,0 +1,104 @@
+        for i in range(1, div):
+            tmp = track.deepcopy()
+            tmp.adjust_time(lambda x : x + i * timesteps)
+            track.notes.extend(tmp.notes)
```
This function seems weird. The last `extend` function duplicates the notes many times.

wu-ming233 replied:
Sorry for the lack of documentation. This function is for the case when the user wants to train short loops which are usually either 4-bars or 8-bars in length per MIDI. For the ones that are 4-bars, we use this function to loop it twice. Note that this `force_length` function is disabled by default when preparing data.

It is true that this would not function as intended if a MIDI has 1 track that's >= 8 bars but some other tracks < 8 bars, as it would loop the tracks < 8 bars. To address this, I implemented a simple check in the newest commit.

aik2mlj commented on polyffusion/train/__init__.py:
```diff
@@ -23,11 +23,13 @@ def __init__(self, params, param_scheduler, output_dir) -> None:
+                    print("New params differ, using new params could break things")
+                    params_file.seek(0)
+                    json.dump(self.params, params_file)
+                    params_file.truncate()
```
Maybe we can leave the old param file unchange since it records the params when training, which should be preserved.

wu-ming233 replied:
Agreed! I implemented that, and allowed users to choose if they want to overwrite the old params file or to save the old params file with a time stamp.

aik2mlj commented on polyffusion/train/__init__.py:
```diff
@@ -40,7 +42,6 @@ def train(self):
             if input("Resume training? (y/n)") != "y":
                 return
         else:
-            output_dir = f"{output_dir}/{datetime.now().strftime('%m-%d_%H%M%S')}"
```
This line is intended. We want the `output_dir` to be able to host multiple trainings with different time stamps.

wu-ming233 replied:
That makes sense. I have reverted the change.

aik2mlj reviewed (changes requested):
Thanks for your great work! I have some concerns as listed below, could you please take your time to review them?

wu-ming233 committed:
Implemented requested changes and an update for prepare_data.py

wu-ming233 commented:
The new commit implements the changes requested above, as well as two minor things:
1. When preparing data from MIDI, we would like to discard the non-melody parts, including drum tracks and melody tracks that have instrument program number >= 113. I've added this option in `prepare_data.py`, and it's enabled by default.
2. Unconditional vanilla ddpm lacked a way to specify which weights to use under the same checkpoint. I've added that, in a similar fashion as what's done in the other ldm models.

Let me know if anything else needs to be changed!

aik2mlj commented:
Looks good to me! Thank you for your efforts.
Now it's only the augmentation range [-6, 7) that needs to be fixed.

wu-ming233 committed:
Reversed collate_fn
